src/iuse.cpp

```diff
@@ -8083,6 +8085,324 @@ std::optional<int> iuse::capture_monster_act( Character *p, item *it, const trip
+            {
+                return pgettext( "short for infinity", "inf" );
+            }
+            return string_format( "%3d", val );
+        };
+        const std::string volume = string_join( display_stat( "", used_volume.value(),
+                                                available_volume.value(),
+                                                to_string ), "" );
+        const std::string ammo = string_join( display_stat( "", required.ammo * heating_effect,
+                                              available_heater,
+                                              to_string ), "" );
+        using stats = inventory_selector::stats;
+        return stats{{
+                {{ _( "Container" ), volume }},
+                {{ _( "Fuel" ), ammo }},
+                {{ _( "Estimated time" ), time }}
+            }};
+    };
+    inventory_multiselector inv_s( *p, preset, _( "ITEMS TO HEAT" ),
+                                   make_raw_stats, /*allow_select_contained=*/true );
+    inv_s.add_character_items( *p );
+    inv_s.add_nearby_items( PICKUP_RANGE );
+    inv_s.set_title( _( "Heat menu" ) );
+    inv_s.set_hint( _( "To heat x items, type a number before selecting." ) );
+    if( inv_s.empty() ) {
+        popup( std::string( _( "You have nothing to heat." ) ), PF_GET_KEY );
+        return std::nullopt;
+    }
+    const drop_locations to_heat = inv_s.execute();
+    if( to_heat.empty() ) {
+        return std::nullopt;
+    }
+    units::volume used_volume = 0_ml;
+    units::mass frozen_weight = 0_gram;
+    units::mass not_frozen_weight = 0_gram;
+    for( const auto &pair : to_heat ) {
+        used_volume +=  pair.first->volume( false, true, pair.second );
+        if( pair.first->has_own_flag( flag_FROZEN ) && !pair.first->has_own_flag( flag_EATEN_COLD ) ) {
+            frozen_weight +=  pair.first->weight( false, false ) * pair.second ;
+        } else {
+            not_frozen_weight +=  pair.first->weight( false, false ) * pair.second;
+        }
+    }
+    heating_requirements required = heating_requirements_for_weight( frozen_weight, not_frozen_weight );
+    if( used_volume > available_volume ) {
+        p->add_msg_if_player( _( "You need more space to contain these items." ) );
+        return std::nullopt;
+    } else if( available_heater < required.ammo * heating_effect ) {
+        p->add_msg_if_player( _( "You need more energy to heat these items." ) );
+        return std::nullopt;
+    }
+    const std::vector<Character *> helpers = p->get_crafting_helpers();
+    const std::size_t helpersize = p->get_num_crafting_helpers( 3 );
+    required.time *= ( 1.0f - ( helpersize / 10.0f ) );
+    for( std::size_t i = 0; i < helpersize; i++ ) {
+        add_msg( m_info, _( "%s helps with this task…" ), helpers[i]->get_name() );
+    }
+    p->assign_activity( heat_activity_actor( to_heat, required ) );
+    if( fire_flag != 1 ) {
+        heater->activation_consume( required.ammo, heater.position(), p );
```

Comment:
This needs to happen at the end of the activity, not at the start.

This prevents the activity being cancelled and still consuming energy.
It's also worth adding a do_turn to the activity to ensure that there still is enough energy/the conditions are right for the heating to continue. If they're not, the activity should be cancelled, with no effect.

Cancellation could result in completing as much heating as was possible for the time elapsed/energy available, but this is not necessary and is complicated for an already big PR.